File: python/core/observability.py
from mcp.types import LoggingMessageNotificationParams


async def logging_callback(params: LoggingMessageNotificationParams) -> None:
    """Prints log messages sent by the server via `ctx.info` / `ctx.warning` / etc."""
    print(f"[server log] {params.data}")


async def progress_callback(
    progress: float, total: float | None, message: str | None
) -> None:
    """Prints progress notifications sent by the server via `ctx.report_progress`."""
    if total is not None:
        percentage = (progress / total) * 100
        label = f"{progress}/{total} ({percentage:.1f}%)"
    else:
        label = f"{progress}"

    print(f"[progress] {message + ' - ' if message else ''}{label}")

File: typescript/src/core/observability.ts
import type { Progress } from "@modelcontextprotocol/sdk/types.js";
import type { LoggingMessageNotification } from "@modelcontextprotocol/sdk/types.js";

// Prints log messages sent by the server via `ctx.info` / `mcp.server.sendLoggingMessage`.
// Equivalent to `logging_callback` in `core/observability.py`.
export function loggingHandler(
  notification: LoggingMessageNotification
): void {
  console.log(`[server log] ${notification.params.data}`);
}

// Prints progress notifications sent by the server via `ctx.reportProgress`.
// Equivalent to `progress_callback` in `core/observability.py`.
export function progressCallback({ progress, total, message }: Progress): void {
  const label =
    total !== undefined
      ? `${progress}/${total} (${((progress / total) * 100).toFixed(1)}%)`
      : `${progress}`;

  console.log(`[progress] ${message ? `${message} - ` : ""}${label}`);
}
